File: Crashkurs Automaten/1 Einführung in Automaten - DFA/1 Getraenkeautomat1und2EuroDFA CS/Getraenkeautomat1und2EuroDFA/DFA.cs
namespace Getraenkeautomat1und2EuroDFA;

public enum Input
{
    _1EURO,
    _2EURO
}

public class DFA
{
    private enum State
    {
        S, Z1, Z2, Z3, ZOVERFLOW
    }

    private State currentState;
    private HashSet<State> endStates;

    private Dictionary<State, Dictionary<Input, State>> transition;

    public DFA()
    {
        endStates = new HashSet<State>{ State.Z3 };
        Reset();

        //                   Eingangszustand -> (Alle Inputs -> Ausgangszustand)
        transition = new Dictionary<State, Dictionary<Input, State>>(5)
        {
            { State.S, new Dictionary<Input, State>
                {
                    { Input._1EURO, State.Z1 },
                    { Input._2EURO, State.Z2 }
                }
            },
            { State.Z1, new Dictionary<Input, State>
                {
                    { Input._1EURO, State.Z2 },
                    { Input._2EURO, State.Z3 }
                }
            },
            { State.Z2, new Dictionary<Input, State>
                {
                    { Input._1EURO, State.Z3 },
                    { Input._2EURO, State.ZOVERFLOW }
                }
            },
            { State.ZOVERFLOW, new Dictionary<Input, State>
                {
                    { Input._1EURO, State.ZOVERFLOW },
                    { Input._2EURO, State.ZOVERFLOW }
                }
            },        
        };
    }

    public void Reset()
    {
        currentState = State.S;

        Console.WriteLine($"Current state: {currentState}");
    }

    public void ProcessInput(object sender, Input input)
    {
        ApplyInput(input);
    }
    
    private void ApplyInput(Input input)
    {
        State newState = transition[currentState][input];
        Console.WriteLine($"Change from {currentState} -> {newState}");
        currentState = newState;
        
        if (endStates.Contains(currentState))
        {
            Console.WriteLine("Endzustand erreicht!");
        }
    }
}

File: Crashkurs Automaten/1 Einführung in Automaten - DFA/1 Getraenkeautomt1und2EuroDFA Java/Getraenkeautomat1und2EuroDFA/src/main/java/at/htldonaustadt/dfa/finitestatemachine/DFA.java
package at.htldonaustadt.dfa.finitestatemachine;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class DFA implements Event{
    private State currentState;
    private Set<State> endStates;

    private Map<State, Map<Input, State>> transition;

    public DFA()
    {
        endStates = Set.of(State.Z3);
        reset();

        // Eingangszustand -> (Alle Inputs -> Ausgangszustand)
        Map<Input, State> mapS  = Map.of(Input.EURO1, State.Z1, Input.EURO2, State.Z2);
        Map<Input, State> mapZ1 = Map.of(Input.EURO1, State.Z2, Input.EURO2, State.Z3);
        Map<Input, State> mapZ2 = Map.of(Input.EURO1, State.Z3, Input.EURO2, State.ZOVERFLOW);
        Map<Input, State> mapZoverflow = Map.of(Input.EURO1, State.ZOVERFLOW, Input.EURO2, State.ZOVERFLOW);

        transition = Map.of(
                State.S, mapS,
                State.Z1, mapZ1,
                State.Z2, mapZ2,
                State.ZOVERFLOW, mapZoverflow
        );
    }

    public void reset()
    {
        currentState = State.S;

        System.out.println("Current state: "+currentState);
    }

    @Override
    public void processInput(Input input) {
        applyInput(input);
    }

    private void applyInput(Input input)
    {
        State newState = transition.get(currentState).get(input);
        System.out.println("Change from "+currentState+" -> "+newState);
        currentState = newState;

        if (endStates.contains(currentState))
        {
            System.out.println("Endzustand erreicht!");
        }
    }
}
